### DISCUSSION TITLE
How to implement 'middleware' in App Router with full node runtime?

### DISCUSSION THREAD

**Discussion by xecrets** [vercel/next.js#54842] [App Router]
In the pages mode, it was possible to inject  your own request handler to do things in `server.js`, based on the documentation here https://nextjs.org/docs/pages/building-your-application/configuring/custom-server . Wrapping the return of `app.getRequestHandler()` with your own handler was easy.

With the App Router, that sample no longer works when deployed. Not sure exactly why, but it just doesn't. It does, sort of, work if you remove the `app.prepare()` bit, but I really can't find any documentation on what that really does and what negative effects it would have to just remove it. I prefer solutions that are documented to work, not just hacks that "seem" to work without understanding why.

With App Router, you can still use the `middleware.js` way of injecting things to handle incoming requests. However... there are some restrictions on what you can do there, since apparently it's limited to the 'edge' runtime, the idea being that these things should be lightweight things done at the edge, like patching some headers etc. But, what I need to do can't be done with the 'edge' runtime.

To get the app working at all with App Router and a custom server, since it's not deployed to Vercel, I did a `create-next-app`, build it with output standalone, and took the `server.js` from there as a starting point.

That `server.js` is based on calling `startServer` with `require("next/dist/server/lib/start-server");`. That works fine as it goes, but I can't seem to find a way to inject my own middleware request handler there, it's all locked up with no hooks that I can find.

At this point I'm kind of stuck.

What I'm actually trying to do, is to track a request with Microsoft Application Insights. The site, to further complicate things, is hosted on IIS with iisnode and it's a request that goes to a static file that doesn't even reach node and Next.js but is handled directly by IIS. The idea would be to reconfigure so the request goes to the Next.js app, and then in the presumed `requestHandler` it would be handled there either, along with the tracking, by way of a redirect to the real static resource, or just by streaming it directly to the client (although I'd rather not do that, since there's a pretty high performance hit then). Another option, which may be what I have to do, is to track not the download, but the click for the download in the client. Other ideas are also welcome. But the question is still valid and remains:

_How to implement 'middleware' in an App Router application with use of the full node runtime, not the limited 'edge' runtime?_

**Comment by cayter** [comment]
Facing the same issue, am wondering if the app router is just having too many issues to tackle custom server at the moment or app router won't be entertaining the self-hosted custom server options in the future. Would really appreciate if we can get some info. Thanks.
